### PR TITLE
Skipping the file if it was not found (the CI is fixed)

### PR DESCRIPTION
**Pull request by DenisRumyantsev**
**Description**: occasionally the `find` function fails with the error `Failed find: ENOENT: no such file or directory`. The reason is that the `fs.lstatSync` method applies to a file that already was deleted. So we should check a file presence and skip it if it does not exist.

**Changelog**: `node/task.ts` — the `find` function — a part of the code where the error occurs was taken out to the separate `_getStats` function. A new field `skipMissingFiles` was added into the `findOptions` parameter. When it is true, the `find` function skips files that are not found during the `_getStats` function execution.

[Related issue](https://github.com/microsoft/azure-pipelines-tasks/issues/14816)

### CODE REVIEW

File: node/task.ts [hunk 1]
```diff
@@ -858,6 +863,45 @@ export function retry(func: Function, args: any[], retryOptions: RetryOptions =
     }
 }
 
+function _getStats (path: string, isFirst: boolean, options: FindOptions): fs.Stats {
```
Comment (max-zaytsev) [inline node/task.ts:866]: It's not clear what is `isFirst `here?

File: node/task.ts [hunk 2]
```diff
@@ -903,49 +947,19 @@ export function find(findPath: string, options?: FindOptions): string[] {
...
-                    if (err.code == 'ENOENT' && options.allowBrokenSymbolicLinks) {
-                        // fallback to lstat (broken symlinks allowed)
-                        stats = fs.lstatSync(item.path);
-                        debug(`  ${item.path} (broken symlink)`);
-                    }
-                    else {
-                        throw err;
-                    }
+            try {
+                stats = _getStats(item.path, !result.length, options);
```
Comment (alexander-smolyakov) [inline node/task.ts:954]: Could you please provide some comments regarding `!result.length` value? Also, I think it makes sense to extract this value as a variable and give it an appropriate name, so the code will be easier to read.

**Review by EzzhevNikita — COMMENTED** [review]
@DenisRumyantsev As far as I understand these changes should fix CI, but looks the issue still persists.

**Comment by DenisRumyantsev** [comment]
> @DenisRumyantsev As far as I understand these changes should fix CI, but looks the issue still persists.

@EzzhevNikita I have already tried to provide similar changes related to [the same issue](https://github.com/microsoft/azure-pipelines-tasks/issues/14816) via [this PR](https://github.com/microsoft/azure-pipelines-task-lib/pull/774), but it broke the CI for Linux and macOS, so I had to [revert](https://github.com/microsoft/azure-pipelines-task-lib/pull/778) it and made a new PR, the current one. For fixing the CI for Windows, there was created [another separate issue](https://github.com/microsoft/build-task-team/issues/1393).

File: node/task.ts [hunk 3]
```diff
@@ -858,6 +863,53 @@ export function retry(func: Function, args: any[], retryOptions: RetryOptions =
     }
 }
 
+/**
+ * Gets info about item stats.
+ *
+ * @param path              a path to the item to be processed.
+ * @param isPathToSearch    true when a path to the item is a path to search.
```
Comment (alexander-smolyakov) [inline node/task.ts:870]:
```suggestion
 * @param isPathToSearch    true when a path to the item equals the path to search.
```

**Review by alexander-smolyakov — APPROVED** [review]
LGTM and please take a look at the suggestion

**Review by EzzhevNikita — APPROVED** [review]

**Review by max-zaytsev — APPROVED** [review]

File: node/task.ts [hunk 4]
```diff
@@ -903,49 +942,25 @@ export function find(findPath: string, options?: FindOptions): string[] {
...
+                const isPathToSearch = !result.length;
+
+                // following all symlinks OR following symlinks for the specified path AND this is the specified path
+                const followSymbolicLinks = options.followSymbolicLinks || options.followSpecifiedSymbolicLink && isPathToSearch;
+
+                // stat the item. The stat info is used further below to determine whether to traverse deeper
+                stats = _getStats(item.path, followSymbolicLinks, options.allowBrokenSymbolicLinks);
+            } catch (err) {
+                if (err.code == 'ENOENT' && options.skipMissingFiles) {
+                    debug(`File "${item.path}" seems to be removed during find operation execution - so skipping it.`);
```
Comment (max-zaytsev) [inline node/task.ts:958]: We have to fix this message since the error here can be not only due to the missing file but due to a broken symlink also

**Review by max-zaytsev — APPROVED** [review]

File: node/task.ts [hunk 5]
```diff
@@ -903,49 +942,25 @@ export function find(findPath: string, options?: FindOptions): string[] {
...
-                        // fallback to lstat (broken symlinks allowed)
-                        stats = fs.lstatSync(item.path);
-                        debug(`  ${item.path} (broken symlink)`);
-                    }
-                    else {
-                        throw err;
-                    }
+            try {
+                // `item.path` equals `findPath` for the first item to be processed, when the `result` array is empty
+                const isPathToSearch = !result.length;
```
Comment (alexander-smolyakov) [inline node/task.ts:949]:
```suggestion
                const isPathToSearch: boolean = !result.length;
```

File: node/task.ts [hunk 6]
```diff
@@ -903,49 +942,25 @@ export function find(findPath: string, options?: FindOptions): string[] {
...
+
+                // following all symlinks OR following symlinks for the specified path AND this is the specified path
+                const followSymbolicLinks = options.followSymbolicLinks || options.followSpecifiedSymbolicLink && isPathToSearch;
+
+                // stat the item. The stat info is used further below to determine whether to traverse deeper
+                stats = _getStats(item.path, followSymbolicLinks, options.allowBrokenSymbolicLinks);
```
Comment (alexander-smolyakov) [inline node/task.ts:955]:
```suggestion
                // following specified symlinks only if current path equals specified path 
                const followSpecifiedSymbolicLink: boolean = options.followSpecifiedSymbolicLink && isPathToSearch;
                
                // following all symlinks or following symlinks for the specified path
                const allowFollowSymbolicLink: boolean = options.followSymbolicLinks || followSpecifiedSymbolicLink;
                
                // stat the item. The stat info is used further below to determine whether to traverse deeper
                stats = _getStats(item.path, allowFollowSymbolicLink, options.allowBrokenSymbolicLinks);
```

File: node/task.ts [hunk 7]
```diff
@@ -858,6 +863,40 @@ export function retry(func: Function, args: any[], retryOptions: RetryOptions =
...
 
+/**
+ * Gets info about item stats.
+ *
+ * @param path                      a path to the item to be processed.
+ * @param followSymbolicLinks       indicates whether to traverse descendants of symbolic link directories.
+ * @param allowBrokenSymbolicLinks  when true, broken symbolic link will not cause an error.
+ * @returns fs.Stats
+ */
+function _getStats (path: string, followSymbolicLinks: boolean, allowBrokenSymbolicLinks: boolean): fs.Stats {
```
Comment (alexander-smolyakov) [inline node/task.ts:874]:
```suggestion
function _getStats (path: string, allowFollowSymbolicLinks: boolean, allowBrokenSymbolicLinks: boolean): fs.Stats {
```

File: node/task.ts [hunk 8]
```diff
@@ -903,49 +942,25 @@ export function find(findPath: string, options?: FindOptions): string[] {
...
+                const isPathToSearch = !result.length;
+
+                // following all symlinks OR following symlinks for the specified path AND this is the specified path
+                const followSymbolicLinks = options.followSymbolicLinks || options.followSpecifiedSymbolicLink && isPathToSearch;
+
+                // stat the item. The stat info is used further below to determine whether to traverse deeper
+                stats = _getStats(item.path, followSymbolicLinks, options.allowBrokenSymbolicLinks);
+            } catch (err) {
+                if (err.code == 'ENOENT' && options.skipMissingFiles) {
+                    warning(`"${item.path}" seems to be a removed file or directory / broken symlink - so skipping it.`);
```
Comment (alexander-smolyakov) [inline node/task.ts:958]: Is it possible to more clearly define the root cause of the error? And display a more accurate error message?

Comment (DenisRumyantsev) [reply node/task.ts:958]: Have no ideas how to catch the difference between a removed file or directory and a broken symlink.
Just as it is impossible to say exactly how and when a file or directory was removed.

**Pull request merged**